polkadot/node/network/collator-protocol/src/collator_side/mod.rs

```diff
@@ -1360,19 +1371,44 @@ where
 		if mode.is_enabled() {
 			if let Some(ref mut implicit_view) = state.implicit_view {
 				implicit_view
-					.activate_leaf(sender, *leaf)
+					.activate_leaf(ctx.sender(), *leaf)
 					.await
 					.map_err(Error::ImplicitViewFetchError)?;
 
 				let allowed_ancestry = implicit_view
 					.known_allowed_relay_parents_under(leaf, state.collating_on)
 					.unwrap_or_default();
 
+				// Get the peers that already reported us this head, but we didn't knew it at this
+				// point.
+				let peers = state
+					.peer_data
+					.iter_mut()
+					.filter_map(|(id, data)| {
+						data.unknown_heads.remove(leaf).map(|_| (id, data.version))
+					})
+					.collect::<Vec<_>>();
+
 				for block_hash in allowed_ancestry {
-					state
+					let per_relay_parent = state
 						.per_relay_parent
 						.entry(*block_hash)
 						.or_insert_with(|| PerRelayParent::new(mode));
+
+					// Announce relevant collations to these peers.
+					for (peer_id, peer_version) in &peers {
+						advertise_collation(
```

Comment:
I understand the race condition, but I still don't get why the collations are not advertised when calling `distribute_collation` adter we built on a known relay parent.

Reply:
This is quite strange for local testing, but in practice it could happen that the collation is not advertised for other reasons. The collator has to be really unlucky.